### PR TITLE
Db/sensirion sps30

### PR DESCRIPTION
**Pull request by danielbernalb**
New Sensirion SPS30 support
Include RSSI function
Texto format by Visual Code
Erase displayEndLine routine because is not used

### CODE REVIEW

File: platformio.ini [hunk 1]
```diff
@@ -15,8 +15,9 @@ build_flags =
     -D SRC_REV=527
 
 # Sensor definition
-   -D HONEYWELL        ; if Honeywell uncomment
+   ;-D HONEYWELL       ; if Honeywell OR Plantower uncomment
    ;-D PANASONIC       ; if Panasonic uncomment
+   -D SENSIRION       ; if Sensirion uncomment
```
Comment (hpsaturn) [inline platformio.ini:20]: The official sensor is PANASONIC for now, please change it.

File: lib/gui-utils/src/GUIUtils.cpp [hunk 2]
```diff
@@ -106,17 +106,6 @@ void GUIUtils::displayBottomLine(String msg)
 #endif 
 }
 
-void GUIUtils::displayEndLine(String msg)
```
Comment (hpsaturn) [inline lib/gui-utils/src/GUIUtils.cpp:109]: What is the reason for delete this method? The end line will be not shown?

File: src/main.cpp [hunk 3]
```diff
@@ -37,40 +38,44 @@
...
 U8G2_SSD1306_128X32_UNIVISION_F_HW_I2C u8g2(U8G2_R0, U8X8_PIN_NONE, 4, 5);
-#else       // display via i2c for D1MINI board
-U8G2_SSD1306_64X48_ER_F_HW_I2C u8g2(U8G2_R0,U8X8_PIN_NONE,U8X8_PIN_NONE,U8X8_PIN_NONE);
+#else         // display via i2c for D1MINI board
+U8G2_SSD1306_64X48_ER_F_HW_I2C u8g2(U8G2_R0, U8X8_PIN_NONE, U8X8_PIN_NONE, U8X8_PIN_NONE);
 #endif
 
 // HPMA115S0 sensor config
 #ifdef WEMOSOLED
-#define HPMA_RX 13   // config for Wemos board & TTGO18650
```
Comment (hpsaturn) [inline src/main.cpp:50]: We have a lot of `format code` changes, please revert those changes and send it in a different PR. Also the  idea is migrate to Google C++ standar like I have in the new firmware of CanAirIO version 2.1, in this repo I configured the vscode settings like [this post](https://medium.com/@zamhuang/vscode-how-to-customize-c-s-coding-style-in-vscode-ad16d87e93bf), but again, it is a another topic different to the main topic of this PR.

File: src/main.h [hunk 4]
```diff
@@ -4,45 +4,53 @@ using namespace std;
...
+#define SENSOR_RETRY 1000 // Sensor read retry
+
+// Sensirion SPS30 sensor
+#define SP30_COMMS SERIALPORT2 // UART OR I2C
+uint8_t ret, error_cnt = 0;
+struct sps_values val;
 
 // Humidity sensor
 Adafruit_AM2320 am2320 = Adafruit_AM2320();
-float humi = 0.0;              // % Relative humidity
```
Comment (hpsaturn) [inline src/main.h:15]: Same situation here, all "format code changes" please in another discussion.

Comment (danielbernalb) [reply src/main.h:15]: ok

File: src/main.cpp (hunk 3 above)
Comment (danielbernalb) [reply src/main.cpp:50]: ok

File: lib/gui-utils/src/GUIUtils.cpp (hunk 2 above)
Comment (danielbernalb) [reply lib/gui-utils/src/GUIUtils.cpp:109]: Is not used now, is doing in other method

File: platformio.ini (hunk 1 above)
Comment (danielbernalb) [reply platformio.ini:20]: ok

File: src/main.cpp [hunk 5]
```diff
@@ -640,13 +731,17 @@ void bleLoop(){
 }
 
 void resetLoop(){
-  if (wifiOn){    
-        if (resetvar == 1199) {      
-        resetvar = 0;
-        delay(45000);   // 45 seconds, reset at 30 seconds
+  if (wifiOn){
```
Comment (hpsaturn) [inline src/main.cpp:734]: Please resolve this issue first in previous PR.

File: src/main.cpp [hunk 6]
```diff
@@ -595,13 +688,11 @@ void bleServerInit(){
   // Create a BLE Characteristic for PM 2.5
   pCharactData = pService->createCharacteristic(
       CHARAC_DATA_UUID,
-      BLECharacteristic::PROPERTY_READ | BLECharacteristic::PROPERTY_NOTIFY
```
Comment (hpsaturn) [inline src/main.cpp:598]: please restore the format code here, on the bluetooth callbacks.

File: src/main.cpp [hunk 7]
```diff
@@ -205,6 +264,40 @@ void sensorLoop()
...
+#else // SENSIRION
+  delay(35); //Delay for sincronization
+  // loop to get data
+  do {
+    ret = sps30.GetValues(&val);
+    // data might not have been ready
+    if (ret == ERR_DATALENGTH){
+      if (error_cnt++ > 3) {
+        ErrtoMess((char *)"-->[E][SPS30] Error during reading values: ", ret);
+        //return(false);
```
Comment (hpsaturn) [inline src/main.cpp:277]: please remove unnecessary comments

File: src/main.cpp [hunk 8]
```diff
@@ -79,20 +82,73 @@ void sensorConfig(){
...
+  sps30.GetErrDescription(r, buf, 80);
+  Serial.println(buf);
+}
+
+void Errorloop(char *mess, uint8_t r){
+  if (r) ErrtoMess(mess, r);
+  else Serial.println(mess);
+  Serial.println(F("Program on hold"));
+  //for(;;) delay(100000);
+  for(;;) delay(500);
```
[the remark this answers is not in the document]
Comment (danielbernalb) [reply src/main.cpp:100]: Yes that's right, corrected.

File: src/main.cpp (hunk 6 above)
Comment (danielbernalb) [reply src/main.cpp:598]: done

File: src/main.cpp (hunk 7 above)
Comment (danielbernalb) [reply src/main.cpp:277]: done

File: src/main.cpp (hunk 5 above)
Comment (danielbernalb) [reply src/main.cpp:734]: commented in previous PR

File: src/main.cpp [hunk 9]
```diff
@@ -205,6 +264,40 @@ void sensorLoop()
...
+      }
+      delay(1000);
+    }
+    // if other error
+    else if (ret != ERR_OK){
+      ErrtoMess((char *)"-->[E][SPS30] Error during reading values: ", ret);
+      //return(false);
+      return;
+    }
+  } while (ret != ERR_OK);
```
[the remark this answers is not in the document]
Comment (danielbernalb) [reply src/main.cpp:282]: is part of the do routine

File: src/main.cpp [hunk 10]
```diff
@@ -157,27 +161,47 @@ void wrongDataState(){
...
+#ifndef SENSIRION
 void saveDataForAverage(unsigned int pm25, unsigned int pm10){
   v25.push_back(pm25);
   v10.push_back(pm10);
 }
+#else
+void saveDataForAverage(unsigned int pm25, unsigned int pm10, float pm25f){
+  v25.push_back(pm25);
+  v10.push_back(pm10);
+  v25f.push_back(pm25f);
```
Comment (hpsaturn) [inline src/main.cpp:190]: this vector, v25f, is a float vector, you tested it with influxdb old schema? or it don't matter?

File: src/main.cpp [hunk 11]
```diff
@@ -504,9 +533,6 @@ void influxDbAddTags(char* tags) {
 }
 
 bool influxDbWrite() {
-  if(apm25 == 0 || apm10 == 0) {
```
Comment (hpsaturn) [inline src/main.cpp:420]: Please sync your branch with the current devel branch, because this lines already committed on previous PR #64 fix

Comment (danielbernalb) [reply src/main.cpp]: donde

File: src/main.cpp (hunk 10 above)
Comment (danielbernalb) [reply src/main.cpp:190]: This is a float vector that carries the float value of PM2.5 of the Sensirion sensor, is sending to the influx and gblabs server via the speed variable in the frame that is float too. I think is provisional for the test that I'm making

Comment (hpsaturn) [reply src/main.cpp:190]: For a test code, you need write a independent code for run test units or similar kind tests code.

File: src/main.cpp [hunk 12]
```diff
@@ -102,79 +178,117 @@ void wrongDataState(){
...
 
-void showValues(int pm25, int pm10)
-{
+#ifndef SENSIRION
+void showValues(int pm25, int pm10){
   gui.displaySensorAverage(apm25); // it was calculated on bleLoop()
   gui.displaySensorData(pm25, pm10, chargeLevel, humi, temp, rssi);
   gui.displayLiveIcon();
   saveDataForAverage(pm25, pm10);
+  WrongSerialData = false;
```
Comment (hpsaturn) [inline src/main.cpp:236]: Why is the reason for have a global variable here?

File: src/main.cpp [hunk 13]
```diff
@@ -425,7 +560,7 @@ bool influxDbWrite() {
 }
 
 void influxDbLoop() {
-  if(v25.size()==0 && wifiOn && cfg.isIfxEnable() && influxDbIsConfigured()){
+  if(v25.size() == 0 && wifiOn && cfg.isApiEnable() && apiIsConfigured() && resetvar != 0 && WrongSerialData == 0){
```
Comment (hpsaturn) [inline src/main.cpp:563]: same here, I don't know why we need a serial variable validation here, for send the payload to the cloud.

Comment (danielbernalb) [reply src/main.cpp]: done, deleted

File: src/main.cpp (hunk 10 above)
Comment (danielbernalb) [reply src/main.cpp]: deleted the pm25f

File: src/main.cpp (hunk 12 above)
Comment (danielbernalb) [reply src/main.cpp:215]: deleted pm25f
WrongSerialData is a flag for correct a bug with the TTGO_TQ that is rest continuously after each power on or reset

**Review by hpsaturn — APPROVED** [review]

**Pull request merged**